File: 0072-edit-distance/0072-edit-distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();
        int[][] dp = new int[m + 1][n + 1];
        for (int i = 0; i <= m; i++) dp[i][0] = i;

        for (int j = 0; j <= n; j++) dp[0][j] = j;
        for(int i = 1; i <= m; i++){
            char c = word2.charAt(i - 1);
            for(int j = 1; j <= n; j++){
                char d = word1.charAt(j - 1);
                if(c == d) dp[i][j] = dp[i - 1][j - 1];
                else dp[i][j] = Math.min(dp[i - 1][j], Math.min(dp[i][j-1], dp[i-1][j-1])) + 1;
            }
        }
        return dp[m][n];
    }
}

File: 0072-edit-distance/0072-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        tab = [[0] * (len(word2) + 1) for _ in range(len(word1) + 1)]
        for i in range(len(word1) + 1):
            tab[i][0] = i
        for j in range(len(word2) + 1):
            tab[0][j] = j
        for i, c in enumerate(word1, 1):
            for j, d in enumerate(word2, 1):
                if c == d:
                    tab[i][j] = tab[i - 1][j - 1]
                else:
                    tab[i][j] = min(tab[i - 1][j - 1], tab[i - 1][j], tab[i][j - 1]) + 1
        return tab[-1][-1]
